### ISSUE TITLE
Investigate functionality to start collateral auctions

### ISSUE THREAD

**Issue by valiafetisov** [sidestream-tech/unified-auctions-ui#363]
### Goal

Get understanding how difficult is it to get all information/rights/etc to start new auctions.

### Context

Starting new auctions or, in maker terms, `barking` on _underwater vaults_ is the functionality that is missing from a complete end-to-end auction flow. Currently, we only provide functionality to facilitate fair market participation in started collateral auctions, and depend on other community members to start auctions. While it is actually profitable by itself and might be required for the newly onboarded collaterals.

### Assets

- `dog.bark` method to liquidate a vault https://github.com/makerdao/dss/blob/fa4f6630afb0624d04a003e920b0d71a00331d98/src/dog.sol#L156-L237
  - `clicp.kick` method to start an auction https://github.com/makerdao/dss/blob/fa4f6630afb0624d04a003e920b0d71a00331d98/src/clip.sol#L220-L266
- Docs https://docs.makerdao.com/smart-contract-modules/dog-and-clipper-detailed-documentation
- Data insights CU and their API https://data.makerdao.network
- Old graphql API for vault analytics https://api.makerdao.com

### Tasks

- [ ] Outline investigation results

**Comment by LukSteib** [comment]
Just wanted to add the following resources here which are nice IMO: 
* https://maker.blockanalitica.com/vaults-at-risk/
  * overview on vaults at risk
* https://maker.blockanalitica.com/simulations/vaults-at-risk/
  * UI to simulate change in OSM price and its effect on vaults at risk

**Comment by KirillDogadin-std** [comment]
> overview on vaults at risk

```sh
curl 'https://papi.blockanalitica.com/maker/vaults-at-risk/' --compressed
```

just sends back the vaults at risk with no auth required, so i guess we could negotiate if it's allowed to borrows this api's functionality in our ui.

The only q becomes wether it only displays vaults at risk or also the ones that are already undercollateralized.

**Comment by KirillDogadin-std** [comment]
__Barking on the auction__

Happens in the following manner:

1. Extracts vault parameters ( stored collateral and the debt) from vat.
2. Make sure that the value stored in the vault is less than the debt, otherwise throws error `Dog/not-unsafe`
3. Check if the trade is possible and overall value that is being auctioned currently does not exceed the limit (which is set in contract)
4. if the value that can still be auctioned is less than the value contained in the vault, partial liquidation is initiated. The check is performed so that the liquidated volume is not to small (regulated by `dust` parameter in collateral)
5. Vat confiscates the vault (the contents of the vault are now belonging to the vat, the debt is belonging to the protocol)
6. Adjust the variable that stores the currently auctioned value 
7. Start the auction
8. emit event

__Detecting underwater and at-risk auctions__

There're services that have been already listed.

1. One of those is using api mentioned at https://github.com/sidestream-tech/unified-auctions-ui/issues/363#issuecomment-1216329903 ,
2. the other one is https://data-api.makerdao.network/redoc#tag/vaults which sadly:
  - only has an endpoint that lists the vault with limited querying capabilities

Apis related questons:

- How many collaterals are supported?
- Ask them if they can provide access to us.
- what is caching time?
- where is src code?
- where is the documentation?
   - https://data-api.makerdao.network/redoc#tag/vaults for data api
- Are the only vaults that are at risk or both at risk and underwater?
- Do we need to support any network except mainnet?
- Do we need OSM based predictions about what happens to the vault at the first version?

Additional verification needed:

- Is there a way to get the list / ... of the vaults from the blockchain directly?

**Comment by zoey-kaiser** [comment]
**Some additional UI based questions:**

- How can we link the collateral auctions view and the new vaults view together? 
  - Indication of how many vaults are at risk on the collateral auction page? Same for number of auctions being shown on the vaults page
  - How can we add links between the vault and collateral auction? In theory a collateral auction should always be linked to a vault. Can we display the vault data on our own page and link the two together?
  - After barking on a underwater vault how can we quickly direct the user from the vault page to the newly created auction?
- The [Blockanalitica](https://maker.blockanalitica.com/vault-types/ETH-B/vaults/28300/) site has many helpful graphs. Do we ideally want to:
  - recreate the same graphs in our UI? (We would have to create estimates on how long this takes and if it is worth it)
  - Link to the Blockanalitica site
    - @LukSteib is blockanalitica a part of MakerDAO or a separate entity?

- How does the frontend barking flow look?
  - User needs to connect a wallet (WalletConnectPanel)
  - User can select the wallet that the incentive will be sent to (New SelectIncentiveWalletPanel?)
  - User needs to execute the transaction (New ExecuteBarkPanel?)

*Very long term UI questions*
- In which ways does it make sense to keep creating new pages? Do we at some point want to combine some pages? 
  - Use this as an example? Is there a way to cleanly display the Vaults on the same page as the auctions?

**Comment by valiafetisov** [comment]
> Happens in the following manner:

In order to review/validate the outlined process, you need to provide a link to the source of the conclusion to every statement in the numbered list.

What about authorisations? What about insensitives? What about market prices, OSM prices, etc?

> - How many collaterals are supported?

Also, how the support is added? Automatic? If manual, what is the process?

You also missed the question about the CORS and possibility allow our UIs to fetch data directly. I suggest to create a separate comment/private meeting issue to collect questions for the other CU.

> - Do we need to support any network except mainnet?
> - Do we need OSM based predictions about what happens to the vault soon-ish at the first version?

Value added and rough effort estimation is needed to answer those kind of questions. Eg:
- Would OSM price/next OSM update time/last update time help user understand what will happen?
- How to get required data?

> Do we ideally want to recreate the same graphs in our UI?

Effort estimation is needed. You as a designer can propose to display graph(s) if it adds a value. Then based on the added value and the effort needed to implement it, the scoping can take place.

> How does the frontend barking flow look?

What about authorisations? What about insensitives? What about market prices, OSM prices, wallet/vat balances, allowances?

**Comment by zoey-kaiser** [comment]
> Effort estimation is needed. You as a designer can propose to display graph(s) if it adds a value. Then based on the added value and the effort needed to implement it, the scoping can take place.

This effort estimation is hard to do right now. Adding graphs in the frontend is not very difficult with a good Vue Package, however getting the data we need to display the graphs is the difficult part. 

BlockAnalitica shows the following information in graphs: 
- Liquidation Ratio
- OSM Price
- Collateralization 
- Protection Score history
- Debt
- Events (Not in Graph, but in a table underneath)

We would have to see if the APIs return the history of the specific vault or only the current value. I would therefore pass the estimation onto @KirillDogadin-std, as he has better insight into the information we get from either the API or Blockchain (depending on what we choose).

> > How does the frontend barking flow look?
As @KirillDogadin-std, stated we currently do not believe that it is needed.

**Comment by KirillDogadin-std** [comment]
> https://github.com/sidestream-tech/unified-auctions-ui/issues/363#issuecomment-1216613866

For now I can tell that there's a history of the vault in the dedicated endpoint in https://data-api.makerdao.network/redoc#tag/vaults/operation/read_vault_history_v1_vaults_vault_history__vault__get . To understand completely we would need to get access to the documentation of this api because there're keys in the response json that are not named clearly enough. So far i see at least partial information coverage there. But i can't tell what's missing because of (yet again) naming.

From the blockchain perpective:  just due to the fact that the liquidation ratio is set manually as far as i understood (https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/spot.sol#L91) it's essentially required to go block by block to fetch the history of the liquidation ratio per collateral. Just from the liquidation ratio it would already mean that we'd have to develop the block-by-block parsing functionality and populate database we operate this. Seems like overly big amount of work.

**Comment by zoey-kaiser** [comment]
> > [#363 (comment)](https://github.com/sidestream-tech/unified-auctions-ui/issues/363#issuecomment-1216613866)
> 
> For now I can tell that there's a history of the vault in the dedicated endpoint in https://data-api.makerdao.network/redoc#tag/vaults/operation/read_vault_history_v1_vaults_vault_history__vault__get . To understand completely we would need to get access to the documentation of this api because there're keys in the response json that are not named clearly enough. So far i see at least partial information coverage there. But i can't tell what's missing because of (yet again) naming.
> 
> From the blockchain perpective: just due to the fact that the liquidation ratio is set manually as far as i understood (https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/spot.sol#L91) it's essentially required to go block by block to fetch the history of the liquidation ratio per collateral. Just from the liquidation ratio it would already mean that we'd have to develop the block-by-block parsing functionality and populate database we operate this. Seems like overly big amount of work.

Thank you for your insight! I would say we settle on the following for now: Let us first create a simple mock to get the initial flow figured out. Once we have a first draft and more insight into what information the APIs can provide, we can discuss the value of adding at least some of the graphs.

**Comment by valiafetisov** [comment]
> Barking on the auction
> Detecting underwater and at-risk auctions

You mean _vaults_, not _auctions_, I suppose? Also, as suggested before, please refrain from using maker terminology if you don't describe particular technical details.

> So addressing osm and market prices seems to be more of a second step since it is more relevant to updating the values of the price in the chain.

I just want you and me to have a broader picture before focusing on the implementation: what makes vaults risky, how often that happens, is it predictable, etc.

> I would therefore pass the estimation onto @KirillDogadin-std,

Before passing it to @KirillDogadin-std, please follow the same process: first answer the question "What data would help user understand what has/will happen?" only then outline how to get this data (or where it's present), which will be the foundation for the estimations.

> Idea is not exactly clear. Are you implying something like this? https://github.com/sidestream-tech/unified-auctions-ui/issues/418

**Comment by KirillDogadin-std** [comment]
> You mean vaults, not auctions, I suppose? ...

Adjusted

> I just want you and me to have a broader picture before focusing on the implementation: what makes vaults risky, how often that happens, is it predictable, etc.

Ok, then 

- risky vaults seems to be more or less off-protocol term that the apis introduce
  - risky vaults are the ones that have percentages of the liquidation ratio and collateralization close to each other. Trivial oversimplified example: 1 BTC is in the vault. Liquidation Ratio is 100%. BTC is worth 100$, debt of the vault is less than 100$ but is close to it (let's say 98$)
- how often it happens is a slippery question because it relies on vault owners' actions and token economics state. I would not speculate on this and make no assumptions here.
- it is predictable in the sense that if we detect the risky vault at one point of time, we can actively check/keep track of its status.

**Comment by KirillDogadin-std** [comment]
Possible useful link for future: contains some terminology clarification and formulas. Good cheatsheet.

**Comment by zoey-kaiser** [comment]
Answer from @KirillDogadin-std in https://github.com/sidestream-tech/unified-auctions-ui/issues/419#issuecomment-1217795353

> Do we only get vaults that are liquidated, ready to be liquidated or something else? (What does "at-risk" mean)

the minimal desired result is to have the vaults that are ready to be liquidated. the best case scenario is to also list vaults that are at risk. vaults at risk are the vaults that might soon become ready to be liquidated. Vaults that are liquidated are not in scope because these vaults are alcready being auctioned and this is covered by our ui.

> What information do we get for every single vault?

- [ ] So far this is to be clarified. For now. basing your guesses on the linked docs makes sense.

> Do we get historical data about the vaults or only the current values?

historical data might be useful at some point, but is not required for the minimal functionality and it's questionable wether we  should use it for displaying vaults at risk. reason for that is the fact that we maybe could just orient ourselves on the current collateralization ratio and the value that is queued in the osm contract.

> The rates module speaks of ...

- [ ] please adjust the comment with links and references to docs and/or contract lines before i answer.

> Can we give the user an estimate

technically yes, but my question would be "how far in advance" would you like to make this prediction? First we have to define the "at risk" margin that we are going to use. e.g. with margin 10% we would consider predictions for all vaults that are within 10% from being underwater.

> Can we give the user a approximate estimate of how much incentive they gain when liquidating a vault?

Yes, the values are stored in the clipper contract of the collateral ([link](https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/clip.sol#L261))

> What values can be used to create this percentage?

we have a collateralization and liquidation ratios. 

in contract terms these are the formulas:

```
Collateralization Ratio = Vat.urn.ink * Vat.ilk.spot * Spot.ilk.mat / (Vat.urn.art * Vat.ilk.rate)
Liquidation ratio = Ilk.mat
```

**Comment by zoey-kaiser** [comment]
> * [ ]  please adjust the comment with links and references to docs and/or contract lines before i answer.

I am referring to [this](https://docs.makerdao.com/smart-contract-modules/rates-module#who-calls-drip). As we already previously established, users need to call `drip` to recalculate certain aspects of the debt involved. (Hence the difference between the ideal rate and the actual rate). 

My question would be can we mock the calculations done by `drip` to show the users a preview of what the ideal rate might be compared to the actual rate. This is something we also do with the collateral auctions (locally recalculating the price drop every `step` by `cut`). 



> Yes, the values are stored in the clipper contract of the collateral ([link](https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/clip.sol#L261))

Based of this I assume, we will have a store where we have a record to stores the `coin` value (incentive amount) per collateral, is this correct? (Or do we embed it into our auction object?)



> in contract terms these are the formulas:

Thank you for the clarification!

> technically yes, but my question would be "how far in advance" would you like to make this prediction? First we have to define the "at risk" margin that we are going to use. e.g. with margin 10% we would consider predictions for all vaults that are within 10% from being underwater.

That is a very good question. Would your proposal be 10% or was it only an example? I currently do not really know what 10% would mean in this case. Could you elaborate your decision making process on why 10% (or another value) might make sense?

**Comment by zoey-kaiser** [comment]
- When a value on been liquidated, we want to show a history of events that lead to the liquidation. How will we fetch these events and the information about a liquidated vault?

**Comment by KirillDogadin-std** [comment]
> My question would be can we mock

yes, we can mock this calculation of this param.

> Based of this I assume, we will have a store

i think its too early to go into technical implementation since we did not cover the whole picture of understanding yet,

> Would your proposal be 10% or was it only an example?

just an example. in this case 10% is collateralization ratio minus liquidation ratio. so 150 cr - 140 lr = 10

**Comment by KirillDogadin-std** [comment]
As promised , here's the summary on how the vaults function


# interacting with vaults.

Here i will try to cover the logic that is under the whole collateralization.

__opening the vault__

Opening the vault essentially means that the wallet owner hands over the collateral to the contract and in exchange receives some amount of DAI. This will mean that the owner is in debt of the vault and the collateral is the guarantee that this debt will be returned one way (repaid) or another (vault is liqudated and collateral is sold). 

- For opening the vault the signer has to call `open` in the cdp [contract](https://github.com/makerdao/dss-cdp-manager/blob/master/src/DssCdpManager.sol)
- To change the amount of the collateral in the vault the owner has to call `frob` and specify the changes in the amount of collateral stored + changes in the debt

__vault parameters__

the vault's basic parameters are the debt of the owner (dai extracted) and the amount of collateral stored inside.
It's always the goal to store more value in the vault than the debt given out. There's a specific margin defined - liquidation ratio. This 
parameter determines the point when the vault's stored value is considered too small relative to the debt. Which leads us to (condition 1):

_condition 1: valult's stored value has to always be greater than the specific parameter determined by the debt and a predefined coefficient_
See (1) below on how to determine if this condition is met

The vault's owner has to pay out stability fees for the received DAI. These fees are accumulated over time in the vault in the form of the additional debt. The stability fee is a fluid parameter that changes with time.

1. Vault charachteristics that are relevant to determine wether condition 1 is met:
	- `rate` - determines the accumulated stability fee.
	- `spot` - determines the maximum value of the debt per unit of collateral.
	- `ink` - stored amount of collateral
	- `art` - initial debt
	- The (condition 1) can  be expressed as `spot * ink > art * rate` where on the left side of ineqality we have the maximum allowed debt and on the right side we have the actual debt.

`rate` is the value that increments over time. For this we would have to perfomr operations on the blockchain. Therefore, the keepers are incentivised to update the rates in order to be able to liquidate vaults and gain profit because updating `rate` essentially updates the accumulated debt value and therefore allows to make some vault into a liquidation target. (detailed in (2) below)

`spot` is something that is determined by the market. The keepers are incentivised to update the value of it to be able to liquidate vaults because this essentially lowers the value (updates the value and it becomes lower) that is stored in the vault. (detailed in (3) below)


2. Updating `rate` is done via calling `drip` function of `jug` contract that accepts the collateral name
	- the side effect is that this calls `fold` function of `vat` contract which 
		- updates the `rate` of the collateral stored in this contract (essentially copies over the value)
		- updates the cumulative debt values for the collateral that are stored in `vat` and `vow` contracts.

3. Updating `spot` parameter is done through `spot` contract via calling `poke` function.
	- This function refers to the price feed and the stored-in-contract liquidation ratio `mat` to calculate the next value.
	- Role of the price feed in this case is taken by so called OSM-Module.
	- There's one osm contract per collateral.
	- One does not have to communicate with OSM directly to update prices.
	- _Extra_: OSM serves as a buffer which feeds the market prices of the collateral into the system with a delay. All OSMs are tracked in the mother contract [here](https://etherscan.io/address/0x76416a4d5190d071bfed309861527431304aa14f) : to look up osm for collateral one converts the string into hex string and uses "Read contract" functionality of etherscan to receive the address. You can use [this](https://web3-type-converter.onbrn.com) to convert the string. E.g. "ETH-A" is "0x4554482d41" and this leads to https://etherscan.io/address/0x81FE72B5A8d1A857d176C3E7d5Bd2679A9B85763 and then find out that the source of the price is https://etherscan.io/address/0x64de91f5a373cd4c28de3600cb34c7c6ce410c85#code contract

__Tracking existing vaults__

The cdp manager implements a linked list that allows to retrieve information about all the existing vaults:
	- collateral name
	- owner address 

by having the collateral name we can get the necessary info from `vat` via `ilks` mapping, each collateral is described as :
  - cumulative initial debt ( without stability fees ) that has been generated by the collateral, `Art`
  - stability fees `rate`
  - maximal debt per unit of collateral `spot`
  - maximal debt `line`
  - minimal possible debt `dust`

by having the vault owner address we can get the necessary info from `vat` via `urns` mapping, each vault is described as:
  - `ink` amount of stored collateral
  - `art` the debt of this vault without stability fees (how much dai was extracted by owner from this vault)

__Detecting vaults at risk__

It is possible to do it via various means. So here i would just draft the brute force approach.

1. initialize index of the vault to be the first index of the linked list stored in the cdp contract 
2. Extract information from `vat` about the vault, calculate the difference  `spot * ink - art * rate`. If it is close to zero but still greater than it, vault is at risk
3. Address the linked list stored in the cdp contract, get the next vault's index.
4. Extract information .... 
5. ....
6. No more elements in the linked list, terminate.

__Sources__

Documentation 

- rates and how they work https://docs.makerdao.com/smart-contract-modules/rates-module
  - updating `rate` https://docs.makerdao.com/smart-contract-modules/rates-module/jug-detailed-documentation
- updating spot https://docs.makerdao.com/smart-contract-modules/core-module/spot-detailed-documentation?q=spot
- listing of vaults https://docs.makerdao.com/smart-contract-modules/proxy-module/cdp-manager-detailed-documentation?q=spot
- vault information https://docs.makerdao.com/smart-contract-modules/core-module/vat-detailed-documentation?q=spot#1.-introduction-summary
- OSM innerworkings https://docs.makerdao.com/smart-contract-modules/oracle-module/oracle-security-module-osm-detailed-documentation?q=spot
- liquidating vaults https://docs.makerdao.com/smart-contract-modules/dog-and-clipper-detailed-documentation?q=spot

Contracts

- https://github.com/makerdao/dss/tree/master/src
- https://github.com/makerdao/dss-cdp-manager/tree/master/src
- https://github.com/makerdao/osm/blob/master/src/osm.sol

**Comment by LukSteib** [comment]
* In the description you are referring to more than the three contracts referenced, right? Can you add all of the related ones (like `spot` , `jug` )

>  The cdp manager implements a linked list that allows to retrieve information about all the existing vaults:
>> - collateral name
>> - owner address

Can you please elaborate on that point a bit more. The `cdpi` provides us with the latest id. However assuming this id to be incremental some of the previous id don't seem to be in the list (see screenshot below). Is this due to vault's that have been liquidated? Is there a way to easily determine the ids of vaults in existence? 

<img width="300" alt="image" src="https://user-images.githubusercontent.com/36229155/185410456-fa2e7a90-4b74-463b-9c46-817db6651b41.png">


> by having the vault owner address we can get the necessary info from vat via urns mapping, each vault is described as:

Was trying to reproduce this step manually via etherscan.  `urns` takes two parameters - can you elaborate what these are?

 > spot - determines the maximum value of the debt per unit of collateral.

Still having a hard time to wrap my head around the concept of `spot`. Can you try to outline the relation between `spot`, the liquidation ratio and the liquidation price for a given vault?

**Comment by KirillDogadin-std** [comment]
> In the description you are referring to more than the three contracts referenced, right?

**Comment by KirillDogadin-std** [comment]
> In the description you are referring to more than the three contracts referenced, right? Can you add all of the related ones (like spot , jug )

i've linked 2 directories that contain contracts. If one is interested in other contracts that are related, they can still use the reference to go look at them. Also specific contracts are already linked in the documentation articles at the top.

> The cdpi provides us with the latest id.

Id is incremental, blockchain numbers are expressed in 16-base which means that `23` is not a valid request. The valid one is `0x23`.

> Is there a way to easily determine the ids of vaults in existence?

i did not see the line of code that deletes the vault's records. 

For finding the non-liquidated vaults the idea in my head that comes up in the form of a very straightforward logic is:

1. get the first id of the vault that belongs to `vat` (aka is confiscated = liquidated)
2. go through the list and write down all the ids that belong to vat.
3. get the numbers from 0 to value incremental counter, remove all numbers that are in the list of liquidated vaults.
4. we now have a list of existing vaults.

> urns takes two parameters - can you elaborate what these are?

collateral symbol and the owner.

- string "ETH-A" is `0x4554482d41`

so something like `urns['0x4554482d41'][<wallet>]` is valid entry

> Still having a hard time to wrap my head around the concept of spot. Can you try to outline the relation between spot, the liquidation ratio and the liquidation price for a given vault?

- Spot is the value that is determined based on the market price: `s = f(m)` , where `s` is spot, `m` is market price and `f()` is the conversion logic.  It defines "how much debt is allowed per unit of collateral". Let's say it's 10 dai per unit.
  - we put in the vault 10 collateral units, extract 80 dai = our initial debt per collateral is 8.
  - time passes, stability fees are accumulated. Our debt grows, becomes 120. At the same time market price of the collateral grows, the spot is redefined and becomes 20.
  - Still 10 collateral in the vault, but our current debt per collateral unit is 12. It's ok though because this value does not exceed 20.
 - ALTERNATIVE VIEW on spot is the following example:
   - if you have 100 ETH in the vault and purpose of the `spot` (doesnt matter how it works here) is to make the protocol count it as you having 80 market-ETH in the vault.
   - Assume 1 ETH = 1000$ on the market, 1 DAI = 1$. Then from the view of the protocol you have 80 000 $ in the vault.
   - If you owe 89 000 DAI, we liquidate the vault.
 - Liquidation ratio is the parameter based on which the `spot` is calculated. The __approximate__ formula is `spot = market price / liquidation ratio`. So if the market price of ETH is 1000$ and the liquidation ratio is 200% then spot is 500 DAI (here 1 DAI = 1 $). 
   - 500 DAI or less of the debt is allowed per 1 ETH

Liquidation price is something that i did not mention in the wiki, i would say it's the price of the collateral contained that would force your vault to be liquidated if the debt exceeds it. it's `spot` * `collateral_amount`.

With this defenition, `spot` is a liquidation price of 1 unit of collateral.

**Comment by LukSteib** [comment]
> Id is incremental, blockchain numbers are expressed in 16-base which means that 23 is not a valid request. The valid one is 0x23

Get your point, just wondering why it seems to be a valid request for other methods. See below. Anyways let's not loose ourselves in these details for now.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/36229155/185437757-50ddab92-4ea2-4725-9b3a-0891361bc3bf.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/36229155/185437834-6832885e-aee7-44b0-a5ad-7577e687b2bc.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/36229155/185437879-fb81db7b-babb-4dc8-9f4c-657554ef1d3e.png">

> get the first id of the vault that belongs to vat (aka is confiscated = liquidated)

Where do you derive that vault belonging to vat = liquidated vault? How are you planning to get these ids?

> collateral symbol and the owner

Providing an `address` received via calling `cdpmanager.owner` and a collateral string received via calling `cdpmanager.ilks` always yields `ink` and `art` of 0 (see example below). Any further hints what I am doing wrong? 

<img width="400" alt="image" src="https://user-images.githubusercontent.com/36229155/185439371-9fd48b6a-d1ba-407f-863a-7ebdc509018c.png">

> Spot is the value that is determined based on the market price: s = f(m)

thx for the further explainers!

**Comment by KirillDogadin-std** [comment]
>  just wondering why it seems to be a valid request for other methods.

ah, my bad, i though that one is obligated to provide hex at all times in etherscan. Looking into the requests the service sends - no, my assumption was not correct. The `23` is perfectly valid. The 0s just indicate that this vault is the only vault that someone has opened. The wiki probably is written in an unclear way which suggests that there's one list that contains all of the vaults. However there's multiple instances of those which are created on the Per-User base.

SO i open 7 vaults, you open 1 vault. My vaults will be in the linked list, your vaults technically too, but the `prev` and `next` values will be `0` which essentially is a single vault.

**Comment by LukSteib** [comment]
> SO i open 7 vaults, you open 1 vault. My vaults will be in the linked list, your vaults technically too, but the prev and next values will be 0 which essentially is a single vault.

Ok makes sense.

> Bad idea from my side. CDP manager does not get affected during confiscation. Then instead we're forced to go over the Bark events and fetch the vaults from there and then determine the owner via cdp manager tools.

Okey. So currently I have the following high level picture in my had. Please re-iterate if flawed. In order to get an overview on currently active vaults we would need to
1. fetch all events from opening a new vault via the appropriate contract
  * -> provides us a long list of all vaults that have been created
2. fetch all events from liquidating an existing vault via the appropriate contract
  * -> allows us to reduce the list from 1. to non liquidated vaults
3. For each vault of the reduced list after 2. fetch collateral type and owner address via the appropriate contract(s)
4. With collateral type and owner address we can fetch the current vault parameters

> are you sure that the vault is not empty?

Ok, got it. Was trying with the wrong parameters. To document step by step what I did in order to get desired vault parameters

1. Searched for an examplary active vault via https://maker.blockanalitica.com/simulations/vaults-at-risk/
  * went with `Vault ID`: 28300
2. Queried `ilks` method of the [DssCdpManager] (https://etherscan.io/address/0x5ef30b9986345249bc32d8928b7ee64de9435e39#readContract) with `28300` as input 
  * obtained `0x4554482d42000000000000000000000000000000000000000000000000000000` as byte32 for `ETH-B`
3. Queried `urns` method of [DssCdpManager] (https://etherscan.io/address/0x5ef30b9986345249bc32d8928b7ee64de9435e39#readContract) with `28300` as input
  * obtained `0x24D86f0DEBe681a34C7bE7E3EaC6F3A6b2517100` as the vault address
4. Queried `urns` method of [Vat](https://etherscan.io/address/0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b#readContract) with output from 2. as first input and output from 3. as second input 
  * obtained `3663707338812318488603` as output for `ink` (ie. vault's amount of collateral) and `3951524264999358736137941` as output for `art`(ie. vault's initial debt)
5. Queried `ilks` method of [Vat](https://etherscan.io/address/0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b#readContract) with output from 2.
  * obtained `1111460595480448191015149282` as `rate` (ie. stability fees?!) and `1205461538461538461538461538461` as `spot` (ie. maximum value of debt per unit of collateral) 
6. Queried `ilks` method of [Spotter](https://etherscan.io/address/0x65c79fcb50ca1594b025960e539ed7a9a6d434a3#readContract) with output from 2. 
   * obtained `1300000000000000000000000000` as `ilk.mat` (ie. liquidation ratio for collateral type)
 7. Computed the collateralisation ratio of the vault at hand via the formula: `Collateralization Ratio = Vat.urn.ink * Vat.ilk.spot * Spot.ilk.mat / (Vat.urn.art * Vat.ilk.rate)`
  * obtained result: `1.3073` -> 130.73 % 
  * similar to what is displayed here: https://maker.blockanalitica.com/vault-types/ETH-B/vaults/28300/

> Liquidation price is something that i did not mention in the wiki, i would say it's the price of the collateral contained that would force your vault to be liquidated if the debt exceeds it. it's spot * collateral_amount.

Calculation of liqudation price seems not correct. In the example used above with all the given params I would expect a liquidation price of $1,558.41 (as shown [here](https://maker.blockanalitica.com/vault-types/ETH-B/vaults/28300/)) but no clue on how to compute

**Comment by LukSteib** [comment]
> actual price that comes in from the osm

Is there a way for us to determine this via contracts? 
I know that querying `ilks` of the [Spotter](https://etherscan.io/address/0x65c79fcb50ca1594b025960e539ed7a9a6d434a3#readContract) contract for a certain collateral provides us with an address for `pip` (ie. the OSM contract for this collateral type). The when querying `src` of the respective OSM contract (e.g. [OSM_STETHUSD](https://etherscan.io/address/0xFe7a2aC0B945f12089aEEB6eCebf4F384D9f043F#readContract)) we get the address of the medianiser contract that is used for this collateral. Does the medianiser contract (e.g. [MedianWSTETHUSD](https://etherscan.io/address/0x2f73b6567b866302e132273f67661fb89b5a66f2#readContract)) allow us to get the actual osm price?

**Comment by KirillDogadin-std** [comment]
`osm.peek()` gives you the current price

**Comment by LukSteib** [comment]
> osm.peek() gives you the current price

Ok, do you know wether this one is protected by auth? Receiving the error below when trying to execute via etherscan: 

`Error: Returned error: execution reverted: OSM/contract-not-whitelisted`

**Comment by KirillDogadin-std** [comment]
indeed, there's the whitelist you have to be on in order to be able to call this function.

so i would say we stick to deriving the price from the known (available) values. which is `spot` and `mat` (liquidation ratio and the safety margin)

**Comment by KirillDogadin-std** [comment]
@LukSteib since you've mentioned that you miss in the understanding regarding the following functions, please first ask see if the previous posts cause any questions.

`drip`

> Updating rate is done via calling drip function of jug contract that accepts the collateral name

> - the side effect is that this calls fold function of vat contract which
>  - updates the rate of the collateral stored in this contract (essentially copies over the value)
>  - updates the cumulative debt values for the collateral that are stored in vat and vow contracts.

`poke`

> The question is again very general and i might not get the purpose of it. In the context of starting the liquidation these do not appear directly relevant since the dog contract relies only on information stored about the vault and the information stored about the collateral. So addressing osm and market prices seems to be more of a second step since it is more relevant to updating the values of the price in the chain. Which seems to be done via poke method in the Spotter [contract](https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/spot.sol#L98)

**Comment by LukSteib** [comment]
On the `drip` and `poke` discussion:

> drip

Follow up questions on that:
* Is there a way to determine in advance what effect calling `drip` would have on the  `rate` of the given collateral type. For a user it's important to know since presumably calling `drip` will incur tx fees 
* Is there a chance for us to determine when `drip` was called for the last time for a given collateral type (e.g. by parsing events of the jug contract)? 

> poke

Probably I am still not capable to put the different pieces together correctly: 
* (as above) Is there a way for us to determine in advance what effect calling `poke` would have on the `spot` of a given collateral? 
  * Especially in the light of your previous comments on OSM price feed having a buffer I am wondering wether `poke` would only need to be called once every X amount of time and otherwise it would be useless way to burn tx fees

---
On the api 

> For now it seems to me that the information provided by the vaults at risk endpoint is sufficient for our goals.

What do you mean by that? Would you recommend using the dedicated endpoint? Or recreating the logic and directly fetch from chain ourselves? 

> the code of the [vaults at risk](https://github.com/makerdao-data/data-api/blob/b798763c55c7b2a3eccd7a803c6741c11781b36a/app/utils/at_risk.py) extraction looks logically correct.

Can you outline what is done there, similar to previous code investigations?

**Comment by KirillDogadin-std** [comment]
> Is there a way to determine in advance what effect calling drip would have on the rate of the given collateral type. For a user it's important to know since presumably calling drip will incur tx fees

yes, as usual we can reimplement on our side the function that computes the value in the contract.

> Is there a chance for us to determine when drip was called for the last time for a given collateral type (e.g. by parsing events of the jug contract)?

yes, this is stored in the `jug.sol` contract in `ilks` mapping in the variable `rho`

> (as above) Is there a way for us to determine in advance what effect calling poke would have on the spot of a given collateral?

yes, the same answer from the above

> Especially in the light of your previous comments on OSM price feed having a buffer I am wondering wether poke would only need to be called once every X amount of time and otherwise it would be useless way to burn tx fees

yes, you could burn money for nothing here potentially.

---

> What do you mean by that?

in the context of the whole post i mean "the endpoint seems like it should work, but we know that it does not since it's still experimental, now i wonder why. If it does not work for 2% of cases, we might just use it".

> Would you recommend using the dedicated endpoint?

not before we find out why it is experimental (aka what cases are there when it does not provide accurate info and what info is not reliable in these cases)

> Or recreating the logic and directly fetch from chain ourselves?

since it's relying on the database that has the caching time, it does not make sense to do this on our side. We would write an api with the same/similar logic, access the same database. I don't see what we win in this case.

> Can you outline what is done there, similar to previous code investigations?

- information about the potential vaults at risk is extracted from the database.
- information about the collaterals' oracles is extracted from the database
- every collateral's oracle is processed and its information is recorded in the local variable via accessing the contract's data
- information about the current and next price is extracted from each oracle's contract and is recorded in the local variable
- parameters of each collateral (`rate`, `spot`, `Art`, `line`, `dust`) is extracted from the vat contract and recorded
- parameters of each collateral (`mat`, `pip`) is extracted from the spot contact and recorded
- information about the vaults that are at risk (`ink`, `art`) is extracted and recorded
- based on above parameters that were recoded for each vault: the next collateralization ratio is computed for every vault
- if the computed future collateralization ratio is going to exceed the liquidation ratio, the vault will end up in the response of the api

**Comment by KirillDogadin-std** [comment]
__Own implemenatation of detecting vaults that are going underwater soon__

1. Fetch all the vaults that exist and store it in the database:
	- Get the value of the latest vault index `cdpi` from [cdp manager contract](https://etherscan.io/address/0x5ef30b9986345249bc32d8928b7ee64de9435e39#code) 
	- Extract owner and Collateral type for each of the vault indices via `owns` and `ilks` mappings in the cdp manager contract
		- new data: `vaultIndex, vaultOwnerAddress, vaultCollateralType`
	- Extract the locked collateral and normalized debt from each vault via `urns` mapping of `Vat` contract:
		- new data: `vaultCollateralAmount, vaultInitialDebtDai`
	- Extract the collateral parameters from the vat contract: stability fees `rate`, price with safety margin `spot`
		- new data: `accumulatedStabilityRate`, `maxDaiPerCollateral`
	- Extract the oracle address from the blockchain via [osmmom contract](https://etherscan.io/address/0x76416a4d5190d071bfed309861527431304aa14f#code)
		- new data: `oracleAddress`
1. Process the database rows where vault can be at risk:
    - agree on the margin that we use to determine wether the vault is at risk. Let's call it `riskCoefficient`
	- extract from the db all the rows which comply with condition`vaultCollateralAmount * maxDaiPerCollateral < accumulatedStabilityRate * vaultInitialDebt * riskCoefficient`
	- refresh the values in these vaults to the latest parameters via blockchain
2. Look up if the vault is going underwater
	- get the current price and the future price of the collateral
		- use `src` variable of the osm oracle contract to find the supplier of data and extract the future price via this supplier: use the `LogMedianPrice` event, [example of the median contract](https://etherscan.io/address/0x83076a2f42dc1925537165045c9fde9a4b71ad97#code .
			- new data: `futureCollateralUnitPrice`
		- use the values to compute the future `spot` value: `futureMaxDaiPerCollateral = futureCollateralUnitPrice / par / mat` where `par` and `mat` are extracted from the Spotter contract.
			- new data: `futureMaxDaiPerCollateral`
- 3. Result:
	- vault is going underwater if `futureMaxDaiPerCollateral * vaultCollateralAmount < accumulatedStabilityRate * vaultInitialDebt`

**Comment by valiafetisov** [comment]
I am not sure if your comment contains distinction between initial loading of data (eg `store it in the database`) and update of that data. Does it? Or do we need to refetch the same thing periodically?

Before I go into details, can you please estimate how many requests will need to be made in order to a) fetch vaults b) update vaults with latest data c) how much time it will take

**Comment by KirillDogadin-std** [comment]
added the segment about refetching.

**Comment by KirillDogadin-std** [comment]
a) fetch vaults - ~30k requests as of now - equal to the amount of the vaults.
b) 4 requests per vault to populate it with the data.

**Comment by valiafetisov** [comment]
Since now it's obvious your proposed solution wouldn't work, can you please outline the alternative technical solution?

**Comment by KirillDogadin-std** [comment]
Well, i don't see how one can avoid using the database for detection of such vaults, therefore from my perspective the problem to fix is keeping the database updated.

For this purpose we could use the same approach as https://github.com/makerdao/pymaker/blob/08821054d009a3b75fd83b248008002391a9c95a/pymaker/auctions.py#L617-L630 implements. There the blocks are filtered out based on the 
contract and then based on the signature of the transaction the function that was called is determined.

Then the move with the database would be to have a service running that just tracks the transactions that are happening in the Vat contract (`frob` for vault manipulation and `grab` for vault liquidation`) 

If this approach sound solid, then i can get more technical description here.

**Comment by valiafetisov** [comment]
> There the blocks are filtered out based on the
contract and then based on the signature of the transaction the function that was called is determined

I am just wondering if the filtering can happen on the provider side and delivered to us via websockets. Can the events be used instead? I see that no events are emitted by the vat contract, but for example for the confiscation the new auction event is fired.

Would listening to `frob` and `grab` solve refetching of all 4 parameters per vault? Would we still need to refetch everything else (eg oracle prices, etc)? How much requests and time we would still need to fetch initial data (or can it be skipped)? What are the alternative for fetching this initial data from the chain (eg, using external public api)?

**Comment by KirillDogadin-std** [comment]
>  Can the events be used instead?

1. with barking and vault liquidation it is possible. here the setup is:
  - call dog contract
  - dog contract has auth in vat to call `grab`
  - dog contract executes the function `grab` and emits the event
2. with vault manipulation no event that definitely allows to know what happens with the vault was found by me.
  - vault manipulation should be done via the cdp manager (according to the documentation)
  - cdp manager calls the `frob` function of vat contract 
  - nor cdp manager, nor vat emit any events during manipulations.

> Would listening to frob and grab solve refetching of all 4 parameters per vault?

i've poked around to get the proposal into more solid condition and reverified:
  - the `vaultCollateralAmount, vaultInitialDebtDai` can be tracked this way.
  - tracking `accumulatedStabilityRate` can be done via events tracking of the `spot` contract. 
  - tracking `maxDaiPerCollateral` has to be done by monitoring calls to `vat.fold`

**Comment by KirillDogadin-std** [comment]
# global feature proposal

- Used sources of information:
	- `dog` - https://github.com/makerdao/dss/blob/master/src/dog.sol
		- functions:
			- `bark` - liquidates the vault
				- events: `Bark`
		- variables:
			- `ilks` - collateral information
			- `Hole` - liquidation limit max
			- `Dirt` - liquidation current
	- `vat` - https://github.com/makerdao/dss/blob/master/src/vat.sol
		- functions:
			- `frob` - adjusts the vault's debt and collateral
			- `fold` - adjusts the minimal debt per collateral
		- variables
			- `ilks` - information about collateral type:
			- `urns` - information about the vaults
	- `jug` - https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/jug.sol
		- functions
			- `drip` - stability fee collection
		- variables
			- `ilks` - known collaterals
	- `spot` - https://github.com/makerdao/dss/blob/60690042965500992490f695cf259256cc94c140/src/spot.sol
		- functions
			+ `poke` - update the maximal debt per collateral
			  - events: `Poke`
		- variables:
			- `ilks` - known collaterals
	- `cdpManager` - https://github.com/makerdao/dss-cdp-manager/blob/master/src/DssCdpManager.sol
		- variables:
			- `urns` - addresses of vaults per vault number
			- `ilks` - collaterals type per vault number
			- `cdpi` - number of vaults
		
- Permissions, Authorizations that will be needed: 
	- None

- Structure
	- Frontend UI
	- Core functionaity to communicate with the blockchain regarding the specific auction
	- API that communicates with the database and provides basic information about the existing vault
	- Database that contains information about the known vaults
	- CRON-like service that periodically looks up what happened on the blockchain and updates the database information.

__Types__

```ts
type CollateralType = 'ETH-A' | 'ETH-B'; // OR infer it from the existing dict OR just set it to be string
interface Collateral {
	type: CollateralType;
	stabilityFeeRate: BigNumber;
	maxDebtPerCollateralUnit: BigNumber;
}
interface VaultCdpContract {
	id: number;
	address: string;
	collateralType: CollateralType;
}
interface VaultVatContract {
	initialDebt: BigNumber;
	collateralAmount: BigNumber;
	collateralType: CollateralType;
}
interface Vault {
	id: number;
	address: string;
	collateralType: CollateralType;
	collateralAmount: BigNumber;
	stabilityFeeRate: BigNumber;
	maxDebtPerCollateralUnit: BigNumber;
	initialDebt: BigNumber;
}
interface LiquidationLimit {
	maximumAmountAuctioned: BigNumber;
	currentlyAuctioned: BigNumber;
}
interface OraclePrices {
	current: BigNumber;
	next: BigNumber;
}
```

__Services__
- FrontendUI requires store functions that communicate with the rest of the stack to receive and send informaton
	- `liquidation` store
		- state:
			- `vaultsAtRisk: number[]`
			- `isLoading: boolean`
			- `liquidationLimitsPerCollateral: Record<CollateralType, LiquidationLimit>`
			- `globalLiquidationLimit: LiquidationLimit`
		- actions:
			- `liquidateVault(index: number)`
			- `getLiquidationLimit(collateralType: CollateralType)`
			- `getGlobalLiquidationLimit()`
	- `vaults` store
		- state
			- `vaults: Record<Vault['id'], Vault[]>`
		- actions:
			- `getVaultsAtRisk()`
- Core functionality
	- `liquidations.ts`
		- `liquidateVault(collateralType: CollateralType, vaultAddress: string, incentiveReceiverAddress: string)`
			- calls `dog.bark()`
		- `fetchGlobalLiquidationLimits() -> LiquidationLimit`
			- calls `dog.Hole()`
			- calls `dog.Dirt()`
		- `fetchLiquidationLimits(collateral: CollateralType) -> LiquidationLimit`
			calls `dog.ilks(collateral)`
		- `fetchLiquidatonEventTransactionNumber`
	- `vaults.ts`
		- `fetchCdpVault(index: number) -> VaultCdpContract`
			calls `cdpManager.urns(index)` to get vault address
			calls `cdpManager.ilks(index)` to get collateral type
		- `fetchVaultsCount() -> number`
			calls `cdpManager.cdpi()`
		- `fetchVatVault(type: CollateralType, vaultAddress: string) -> {vault, collateral}: { vault: VaultVatContract, collateral: Collateral }`
			- calls `vat.urns(type, vaultAddress)` to get the collateral amount and initial debt
			- calls `vat.ilks(type, vaultAddress)` to get the collateral configuration
		- `fetchVault(index: number) -> Vault`
			- calls `fetchCdpVault` and `fetchVatVault` mentioned above
		- `isAtRisk(v: Vault, marginCoefficient = 1.2) => v.maxDebtPerCollateralUnit * v.collateralAmount < v.initialDebt * stabilityFeeRate * marginCoefficient`
- Api endpoints
	- GET `/vaults_at_risk/?offset=xx&limit=xx`
		- queries the database for indexes of the vaults that are at risk
		- returns the vault list
	- GET `/vaults/:id`
		- queries the database for vault with the index
		- returns the vault 
- Vault update service
	- env variable:
		- `DEV` - path to recovery file. If file does not exist - generates 100 random numbers and populates the vault table with the ids from this list, caches the contents into file. Recovers from file if the path is set.
		- `DO_REFETCH_ALL_VAULTS` that triggers the complete refetch of the vaults. Does not trigger (throws error) if the `DEV` envvar is set.
		- `UPDATE_FREQUENCY` - number of seconds
	- runs `fetchVault` from the definition above for every index less than `fetchVaultsCount` return if we need to fetch vaults.
	- every `UPDATE_FREQUENCY` seconds runs the update function
	- update:
		- https://docs.ethers.io/v5/api/contract/contract/#Contract-on use the subsription on `dog` contract to react to `Bark` events and set vaults to be liquidated in the db.
		- set listener to `spot` contract and listen to `Poke` to update the values in the db of `maxDebtPerCollateralUnit`
		- get the block number of the latest update from the database.
		- for every block after the queried one parse it:
			- `provider.getBlockWithTransactions` to get block
			- find the signatures of the blockchain methods that are of interest. E.g.: https://docs.ethers.io/v5/api/utils/hashing/#utils-id
			- functions to track in `vat` contract: `fold`, `frob`
				- `fold` influences the adjustment of vault's initial debt and collateral amount
				- `frob` influences the adjustment of the collateral stability fee rate.
- Database
	- Vault table:
		- id: number;
		- address: string;
		- initialDebt: BigNumber;
		- collateralAmount: BigNumber;
		- FK to collateral table
	- Collaterals table
		columns as in `Collateral` type declared above
	- Latest block table:
		- single row with the number of block that the `Vault` table represents
	- ORM: TypeOrm

**Comment by LukSteib** [comment]
Thx for the proposal.

Based on the previous discussions I am still lacking an answer on the question:
How feasible is this approach?
Can you provide an opinion on that?

---
Some scope and volume related input:

I've spent some time looking into solutions that display vaults (especially: https://tracker-vaults.makerdao.network/). Guiding question: How to limit scope of requests we potentially would need to make.
* There is this clear distinction between `total vaults` ~30k and `active vaults` ~2.7k. So only focusing on the latter figure would already quite significantly reduce the volume we are talking about. 
  * related question to @KirillDogadin-std: Do you know how it is determined wether a vault is active or not? (see related repo: https://github.com/makerdao-data/vaults-tracker
* Another (less optimal) approach: Slicing by collateral types and focus on lower volume collaterals first (e.g. `WSTETH-B` with 65 `active vaults` vs. `ETH-A` with 1209 `active vaults`

**Comment by valiafetisov** [comment]
> How feasible is this approach?
> Can you provide an opinion on that?

For my understanding, were those two questions already answered somewhere else?

> Types

As discussed in the daily, 
- [ ] Please extend instead of repeating the types
- [ ] Please choose better namings for the `*Contract` types, since provided values are not the contract
- [ ] Please provide a single end type of the vault that includes everything that needs to be displayed according to the mock (including)
  - [ ] Please make sure it includes everything that is displayed in the mock (eg `proximityToTheLiquidation`, `nextPriceChange`, etc)
  - [ ] Please make sure it includes transaction fees, combines transaction fees, etc
  - [ ] It might make sense to add `state` eg `loaded`, `liquidated`, etc
- [ ] Please provide a single end "context" type per collateral (or merge everything into the vault type above) eg (`LiquidationLimit` is per collateral it can be delivered per collateral or can be delivered inside the valut type, depends on the implementation
- [ ] Please submit types-only PR where we will further discuss the types

**Comment by KirillDogadin-std** [comment]
> For my understanding, were those two questions already answered somewhere else?

sry, forgot to document the verbal discussion;

yes & the answer is it is feasible given that we have the access to the outside database which we use to "copy over" (The mentioned data-api for example) from periodically OR if we do not mind from time to time use up a lot of requests and time to refresh the database state.

**Comment by valiafetisov** [comment]
> yes & the answer is it is feasible given that we have the access to the outside database

- What would be the number of requests in that case? How much time it will need to `sync` to the latest info?
- What is the exact API endpoint we will be dependent upon?
- Can you please outline the architecture: as far as I understand this feature requires the database and a new service that is run constantly? Or does it only updates cache during requests to the endpoints?
- Can we still aim for the frontend to be able to work without this service and the database (in case no vaults liquidation feature is needed or it is down) – for example, when opening /vaults?id=100 we might do all requests frontend-side so in case of market crash even if our service is overloaded, DICU endpoint is overloaded (or someone is just using our UI with their own RPC node) we will receive latest data directly from the chain. Is it feasible? If you agree, we might remove `/vaults/:id` endpoint from above

> * `DEV` - path to recovery file. If file does not exist - generates 100 random numbers and populates the vault table with the ids from this list, caches the contents into file. Recovers from file if the path is set.

Probably a better name is `VAULTS_RECOVERY_FILE`
- What is the format of the file? Should we use `sqlite` as the database and the recovery file as well?
- What would production also need this in case it restarts eg during deployment?
- Where do we fetch this file during development? Is there some kind of endpoint on staging?

> * `DO_REFETCH_ALL_VAULTS` that triggers the complete refetch of the vaults. Does not trigger (throws error) if the `DEV` envvar is set.

I would argue that refetching should be the default logic until some `isDevelopment` flag is set.

> * `UPDATE_FREQUENCY` - number of seconds

Better naming is `VAULTS_REFETCH_INTERVAL_SECONDS` – env variable names for amounts should always include denomination units. Btw, are we actually planning to refetch vaults since I thought we would use websocket delivery to receive events about changes?

**Comment by valiafetisov** [comment]
> The explainer here is not clear to me. Let me make a statement and let's see if it answers your question: /vaults/:id would not be absolutely critical endpoint to have if the frontend already knows the ids it's interested in and wants to update the store state with the latest information.

The question is: in case the user knows the vault id (eg is coming from the twitter that just announced that vault #100 is in liquidatable state) can we fetch all related data directly from the chain on page load instead of relying on our and DICU dbs: a) avoiding the need for `/vaults/:id` endpoint or b) doing this in case endpoint throws an error/is outdated. In any case, we should also add `lastSyncedAt` to each `Vault` in the database

> Where do we fetch this file during development? Is there some kind of endpoint on staging?

During the deployment, do we need to wait for 5 minutes every time we start a dev server and do few thousands requests? Can you propose a mechanism to avoid it that is not an overkill. I imagine an endpoint that returns sqlite file, so that it can be easily curled from the development of even synced under the hood without extra command

```
curl https://unified-auctions.makerdao.com/api/vaults.sqlite --output vaults.sqlite
```

> Does not trigger (throws error) if the `VAULTS_RECOVERY_FILE` envvar is NOT set.

I imagine that it makes sense to have default value for the `VAULTS_RECOVERY_FILE` env var.

**Comment by KirillDogadin-std** [comment]
> The question is

yes, we can. adding the column to the db model

> During the deployment, do we need to wait for 5 minutes every time we start a dev server and do few thousands requests? Can you propose a mechanism to avoid it that is not an overkill. I imagine an endpoint that returns sqlite file, so that it can be easily curled from the development of even synced under the hood without extra command

If i understand correctly, you're talking about the development flow where each time when we start the app it's mandatory to throw thousands of requests. The initial proposal is to only throw 1 request, get ~100 vaults and that's it. Then when the local file exists it can be mounted to the image and not refetched every time.

**Comment by valiafetisov** [comment]
> yes, we can. adding the column to the db model

So option `b)` it is? Can you reflect it in the proposal?

> If i understand correctly, you're talking about the development flow where each time when we start the app it's mandatory to throw thousands of requests

I am talking about a way to have complete experience without needing to refetch everything. Fetching only 100 vaults is not ideal in case you want it to kickstart and proceed with vault liquidations (eg the `bot` is privately run by default will start without a database by a person who just cloned it or will start with outdated database in case it wasn't run until a crash) – all those are valid use-cases.

**Comment by KirillDogadin-std** [comment]
> So option b) it is? Can you reflect it in the proposal?

in fact i was leaning towards option a) where the specific vault is instead just fetched from the blockchain. And this is reflected accordingly already.

> I imagine an endpoint that returns sqlite file, so that it can be easily curled from the development of even synced under the hood without extra command

so then to get this proposal straight:
  - the endpoint serves the latest backup file
  - the mechanism that starts the api server invokes the `curl` command before starting the main functionality.

**Comment by valiafetisov** [comment]
> > I imagine an endpoint that returns sqlite file, so that it can be easily curled from the development of even synced under the hood without extra command
> 
> so then to get this proposal straight before i add it

Well, those were my thoughts on the subject. I don't know if sqlite can be served directly, if it need to be exported first into another file or if it should be exported on-the-fly (ie streamed) and then cached on the proxy level. This can be answered by reading the docs/stack overflow. I also don't know if there is a better/more feasible solution.

> the mechanism that starts the api server invokes the curl command before starting the main functionality

I am not sure if this mechanism should be part of the server or it should be just mentioned in the readme and happy to hear your argumentation. I am sure it need to be optional, since the data in sqlite is not guaranteed to be correct (to the level the blockchain guarantees it) and advanced users wouldn't want to depend on it.

Regarding the updated proposal:

- I don't see the reason why do you need to have separate `liquidation` store. If there are vaults stored already, you can derive `vaultsAtRisk` from the `proximity to liquidation` of each vault. It is especially important in case we allow independent fetching of vaults by their id
- `Limits`-related functionality might also be outdated after the discussion on types in #446 
- `vaults` store is currently missing action to `getVaultById()`
- `Core functionality` still mentions old types
- `Database` might also be outdated after the discussion on types in #446

**Comment by KirillDogadin-std** [comment]
> I am not sure if this mechanism should be part of the server or it should be just mentioned in the readme and happy to hear your argumentation.

I would actually vouch for leaving it up to the user to download the backup manually because:
  - it seems like bad manners to store anything on the user's hard drive implicitly as not an installation app
  - there will have to be no logic written that covers for various scenarios, e.g.:
    - dev env var is set and backup path points to non-existant file -> do we download anything?
    - dev env var is not set and the file is present -> do we download anything?
  - instead of covering for scenarios it's way easier to mention that in the readme and provide tooling that recovers db from the file in path.

> Core functionality still mentions old types

fixed

> Database might also be outdated after the discussion on types

updated

> I don't see the reason why do you need to have separate liquidation store

removed

___

Currently the db contains minimal required information to overview a vault and determine wether it's underwater. all the other parameters are removed since it does not seem to be the db's responsibility to handle anything but allowing to find liquidatable vaults.

**Comment by valiafetisov** [comment]
> I would actually vouch for leaving it up to the user to download the backup manually because:

Agreed, let's provide working curl command in the readme of the new service


> fixed
> `getLiquidationLimit(collateralType: CollateralType)`
> `getGlobalLiquidationLimit()`

I thought we agreed in #446 to receive this information together with each vault?

> collateral symbol (id)

At least in our codebase `collateralSymbol` is for example `ETH`, while `collateralType` is `ETH-A` – I suppose you meant `collateralType` here since different types have different risk parameters. By `(id)` I suppose that it's a primary column

> Currently the db contains minimal required information to overview a vault and determine wether it's underwater

So the core functionality will have a function `getVaultsAtRisk` to: 1) fetch all rows from the collateral table 2) get OSM prices etc from the chain 3) _query_ vaults at risk 4) build `VaultTransaction` type that is returned to the frontend/bot

- I think the important part is that database should be able to "query" vaults at risk without putting them all into memory (which otherwise becomes a 5-10K loop that runs `isAtRisk` on every call) 
- I suppose this "enrichment" of the database happens inside the endpoint so it can be cached on the reverse-proxy level or do we enrich it on each connected client?
- Code organisation-wise I suppose that the new service folder will only contain functions to work with DB and the API, the rest is imported from the core

**Comment by KirillDogadin-std** [comment]
> I think the important part is that database should be able to "query" vaults at risk without putting them all into memory (which otherwise becomes a 5-10K loop that runs isAtRisk on every call)

Given that requirement would indexing help with the number of loops run?

> I suppose this "enrichment" of the database happens inside the endpoint so it can be cached on the reverse-proxy level or do we enrich it on each connected client?

could you please elaborate on what "enrichment" exactly means?

> Code organisation-wise I suppose that the new service folder will only contain functions to work with DB and the API, the rest is imported from the core

yes indeed

**Comment by valiafetisov** [comment]
> Given that requirement would indexing help with the number of loops run?

Indexing would only help for faster fetching of the specific rows and only for queries using indexed columns. I meant that because you have `isAtRisk` function in your proposal you expect to fetch ALL vaults from database, then iterate them over in memory, then filter. Instead, the proposal is to build a query and then directly fetch vaults that are at risk at the moment using query. For example, you have `currentUnitPrice` and `nextUnitPrice`, you determine the lowest `riskyUnitPrice` from it, then you fetch from the database ONLY vaults that are close to the lowest price (or using any other field).

> could you please elaborate on what "enrichment" exactly means?

The process of turning minimal info about the vault stored in the DB into the final `VaultTransaction` type

> 2) get OSM prices etc from the chain
> 3) query vaults at risk
> 4) build `VaultTransaction` type that is returned to the frontend/bot

**Comment by KirillDogadin-std** [comment]
> Indexing would only help for faster fetching of the specific rows and only for queries using indexed columns. I meant that because you have isAtRisk function in your proposal you expect to fetch ALL vaults from database, then iterate them over in memory, then filter. Instead, the proposal is to build a query and then directly fetch vaults that are at risk at the moment using query. For example, you have currentUnitPrice and nextUnitPrice, you determine the lowest riskyUnitPrice from it, then you fetch from the database ONLY vaults that are close to the lowest price (or using any other field).

miscommunication from my side probably.

`isAtRisk` initially was added to cover the case when the vault was fetched from blockchain by index and then it's necessary to figure out if it's at risk.
Since now we had other discussions that lead to setting the `state` on enrichment this function is not relevant anymore and thus can be removed. Updated the proposal.

> I suppose this "enrichment" of the database happens inside the endpoint so it can be cached on the reverse-proxy level or do we enrich it on each connected client?

my vision is to enrich this on each connected client (store calls function from the core that uses blockchain to enrich / update the data) - this guarantees that the client at all times gets fresh data directly from the chain. Also this reduces the dependancy on the api.

**Comment by valiafetisov** [comment]
> my vision is to enrich this on each connected client

If you would already have the data to determine if vault is at risk or not, why would you not add it to the data contained in the database? Why do same things `2+n` times (where `n` is the number of connected clients)?

What would be the number of requests in case we need to enrich say 1000 vaults?

**Comment by KirillDogadin-std** [comment]
Database is technically not a source of truth. I see the point that we can store relevant info there right away and then the users of the ui will not have to address blockchain because they will already have the info from the api. But my concern here becomes: the users do not get the data from the blockchain = they do not get data from a truthful source. Therefore the idea here is to have the db and api for the sole purpose of providing the direction and let the users know in which vault's they're interested but not the complete info about the vault: from the user's perspective it's a potential lie. 

> What would be the number of requests in case we need to enrich say 1000 vaults?

i'm exlcuding the oracle prices fetches, incentive fetches and the limit fetches in this answer: 3 requests per vault = 3000.

if in case of market crash this is extremely likely that there will be thousand(s) of liquidatable vaults then not providing this information along with api indeed is a bad idea. Is this case extremely likely?

TL;DR i'm not a strong proponent of getting everything from the blockchain - we can put as much info as possible into the db if it's ok from the d-app "morales" perspective.

**Comment by valiafetisov** [comment]
> i'm exlcuding the oracle prices fetches, incentive fetches and the limit fetches

I would include all of them to the estimation

> 3000

So that will be the number of requests made will be 3000 on the server, 3000 on the client for the first client + 3000 on the client for each next connected client? Vs 3000 on the server + 0 on the client?

> Is this case extremely likely?

This is the main purpose of the UI and the keeper we're building: to provide solid backstop solution in the event of a market crash.

My proposal stays the same: provide as much information as possible with the API response, then cache it, but fetch information directly from the chain in case user already opened specific vault (from the table or directly using url)

One other point I wanted to bring before we conclude this: `CRVV1ETHSTETH` vaults ids are not numbers:
- https://tracker-vaults.makerdao.network/collateral/CRVV1ETHSTETH-A
- https://tracker-vaults.makerdao.network/collateral/ETH-A (and all other vault types to my knowledge)

**Comment by KirillDogadin-std** [comment]
> One other point I wanted to bring before we conclude this: CRVV1ETHSTETH vaults ids are not numbers:

inserting this values into the `ilks` reader on etherscan https://etherscan.io/address/0x5ef30b9986345249bc32d8928b7ee64de9435e39#readContract returns empty result. In the vaults tracker the owner is not listed. It seems to me this data in the tracker is not really valid... do you have any knowledge about why this is the case currently?

> My proposal stays the same

alright, then i adjust the db model

**Comment by valiafetisov** [comment]
> It seems to me this data in the tracker is not really valid

It is certainly valid

> do you have any knowledge about why this is the case currently?

Maybe they are processed using different contracts? You can quickly investigate by reading [`clip.kick` source code of this collateral type](https://etherscan.io/address/0x1926862f899410bfc19fefb8a3c69c7aed22463a#code). We might put it out of scope for now if it's something totally different.

**Comment by valiafetisov** [comment]
The feature proposal can be concluded. This issue should be reopened and updated in case we receive more info or contradicting info from the DICU
